---
title: "Creating a basis using cupcake"
author: "Olly Burren"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

The function of this package is to facillitate the building of a set of ordered *basis* vectors from a collection of case/control genome-wide association study summary (GWAS) $\beta$, or $\log(\text{Odds Ratio})$ statistsics. The building of ordered sets of basis vectors for analysing GWAS genotype data is not new and it's utility in adjusting for confounding due to population structure is well described. This package seeks to address a different set of questions as to the overall relationship between a set of related (or unrelated) traits for which genotype data may not be available. This follows a logical workflow as follows:-

1. Identify a set of traits of interest from which to construct the basis and collect relevant summary statistics.
2. Idenification of the maximal set of SNPs that are included in all source traits.
3. Odds ratio (OR) alignment, such that the effect allele for all the traits is standardised.
4. *Basis construction.*
5. *Projecting new traits onto the basis.*

This package is concerned with steps 4 and 5, and whilst the other steps are non-trivial this package assumes access to a set ($n\gt1$) of  OR aligned summary statistics (see section on file formats below as to what is required).

## Installing cupcake

## Installing COGS

You will need to have the devtool R package installed
```{r,eval=FALSE}
library(devtools)

install_github("ollyburren/cupcake")
```


## Input files

The package expects two manifest files, one describing the SNPs to include in the basis and the other describing the traits. The package makes the choice to define a `pid` or primary identify for each SNP, that is a portmanteau of it's chromomsome and position. For example, given a reference genome assembly of GRCh37, `rs2476601` which has a position on  chromosome 1 of 114,377,568 has a pid of `1:114377568`. Whilst variants with the same position but differing alleles may exist, the fidelity of any genotyping calls of such a class of variants means that, in practice they are best removed before basis construction.

### SNP Manifest

Each row of this file describes a non redundant (see above) SNP to include in the basis.

1. pid - See above.
2. ref_a1 - Allele 1 (A/G/C/T/I/D).
3. ref_a2 - Allele 2 (A/G/C/T/I/D) - the effect allele.
4. ref_a1.af - The allele frequency of ref_a1 (estimated from a suitable reference population).
5. ld.block - A number grouping a block of SNPs in high LD (see note).

e.g.

| pid       | ref_a1 | ref_a2 | ref_a1.af         | ld.block |
|-----------|--------|--------|-------------------|----------|
| 1:1118275 | G      | A      | 0.95729537366548  | 2        |
| 1:1120431 | G      | A      | 0.956702253855279 | 2        |
| 1:1218086 | G      | A      | 0.924436536180308 | 2        |
| 1:1335790 | A      | G      | 0.820026100367778 | 2        |
| 1:1497824 | G      | A      | 0.875919335705813 | 2        |


### Trait Manifest

Each row of this file describes a trait to either include in basis creation or to project.

1. include - `Y` indicates that a trait should be considered anything else will be excluded.
2. trait - Human readable name for the trait e.g. Juvenile Idiopathic Arthritis.
3. cases - Number of cases in the study.
4. controls - Number of controls in the study.
5. basis_trait - `1` indicates that a trait should be used to create a basis rather than for projection.
6. file - Name of the summary statistic source file (see below) to use e.g. JIA.tab.

e.g.

| include | trait                      | cases | controls | basis_trait | file    |
|---------|----------------------------|-------|----------|-------------|---------|
| Y       | Autoimmume Thyroid Disease | 2733  | 9364     | 1           | ATD.tab |
| Y       | Multiple sclerosis         | 14498 | 24091    | 1           | MS.tab  |
| Y       | Primary billiary cirrhosis | 2861  | 8514     | 1           | PBC.tab |

## GWAS source file

Each row of this file describes the effect and signficance for a set of alleles for a trait which can be derived from the set of GWAS summary statistics for a trait. Effect alleles are expected to match the snp manifest and are duplicated to allow checking. The file should contain either a matching or superset of SNPs described in the SNP manifest.

1. pid - as for the SNP manifest.
2. a1 - Allele 1  - should match SNP manifest.
3. a2 - Allele 2 (effect allele)
4. or - Odds ration with respect to a2.
5. p.value

e.g.

| pid          | a1 | a2 | or    | p.value |
|--------------|----|----|-------|---------|
| 10:100134036 | A  | G  | 0.941 | 0.2727  |
| 10:100179274 | G  | A  | 0.914 | 0.1358  |
| 10:100286047 | C  | A  | 1.05  | 0.4344  |
| 10:100286138 | G  | C  | 1.04  | 0.493   |
| 10:10028820  | G  | A  | 0.918 | 0.1125  |

## Loading in data

The loading of data is handled through `get_gwas_data` routine like so

```{r}
library(devtools)
load_all('~/git/cupcake')
SNP_MANIFEST_FILE <- '../inst/extdata/snp_manifest.tab'
TRAIT_MANIFEST_FILE <- '../inst/extdata/trait_manifest.tab'
DATA_DIR <- '../inst/extdata/gwas_data/'
basis.DT<-get_gwas_data(TRAIT_MANIFEST_FILE,SNP_MANIFEST_FILE,DATA_DIR)
```

The `data.table` object `basis.DT` contains the loaded data required for the next step of calculating the shrinkage values to appy to the $\beta$ of all traits.

```{r}
library(magrittr)
head(basis.DT,n=2) %>% knitr::kable()
```

Basically a merge between the metadata for the two basis traits in the example trait manifest file and the meta data on each SNP, contained in the snp manifest file.

## computing shrinkage

Next we use the `data.table` created to compute shrinkages like so.

```{r}
shrink.DT<-compute_shrinkage_metrics(basis.DT)
```

This computes various shrinkages, for this example we will take forward `ws_emp`, the types of different shrinkages and their names are are explained later in this vignette.

## creating the basis

The final step is to create a basis from the two input traits, firstly we apply shrinkages and line up variants across studies using `create_ds_matrix`. Next we add a null or `control` study in order to anchor the basis. In practice this means adding a 

```{r}
basis.mat.emp <- create_ds_matrix(basis.DT,shrink.DT,'ws_emp')
## need to add control where beta is zero
basis.mat.emp<-rbind(basis.mat.emp,control=rep(0,ncol(basis.mat.emp)))
pc.emp <- prcomp(basis.mat.emp,center=TRUE,scale=FALSE)
```

Visualising

```{r,fig.width = 5}
plot.DT <- data.table(trait=rownames(pc.emp$x),pc.emp$x)
xlim <- with(plot.DT,c(min(PC1),max(PC1)) * 1.2)
ylim <- with(plot.DT,c(min(PC2),max(PC2)) * 1.2)
with(plot.DT,plot(PC1,PC2,type='n',xlim=xlim,ylim=ylim))
with(plot.DT,text(PC1,PC2,labels=trait, cex= 1, offset = 20))
```


## projecting data onto the basis

```{r}
proj.DT<-get_gwas_data(TRAIT_MANIFEST_FILE,SNP_MANIFEST_FILE,DATA_DIR,trait_list='trait3')
proj.mat.emp <- create_ds_matrix(proj.DT,shrink.DT,'ws_emp')
pred <- predict(pc.emp,proj.mat.emp)
```

Visualising projected data

```{r,fig.width = 5}
## plotting code
plot2.DT <- rbind(data.table(trait=rownames(pred),pred),plot.DT)
plot2.DT[,cat:=ifelse(trait=='trait3','projected','basis')]
xlim <- with(plot2.DT,c(min(PC1),max(PC1)) * 1.2)
ylim <- with(plot2.DT,c(min(PC2),max(PC2)) * 1.2)
cols <- c('red',rep('black',3))
with(plot2.DT,plot(PC1,PC2,col=cols,type='n',xlim=xlim,ylim=ylim))
with(plot2.DT,text(PC1,PC2,labels=trait, cex= 1,col=cols))
```


