<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Projecting your data onto an exisiting basis • cupcake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Projecting your data onto an exisiting basis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cupcake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/create-basis.html">Creating a basis using cupcake</a>
    </li>
    <li>
      <a href="../articles/project-basis.html">Projecting your data onto an exisiting basis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Projecting your data onto an exisiting basis</h1>
                        <h4 class="author">Olly Burren</h4>
            
            <h4 class="date">2020-01-08</h4>
      
      
      <div class="hidden name"><code>project-basis.Rmd</code></div>

    </div>

    
    
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span>opts_chunk<span class="op">$</span><span class="kw">set</span>(<span class="dt">echo =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(cupcake)</code></pre></div>
<pre><code>## Loading required package: data.table</code></pre>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This vignette shows you how to project summary GWAS statistics from your study onto a prexisting basis. As part of the paper we created a basis from 13 immune-mediated diseases, this basis is sparse and contains only a few 100 SNPs. In fact the loading matrix is so small that we have bundled it with this package allowing you to conduct your own analysis.</p>
<p>To illustrate this process we will use a cut down set of variants for summary statistics from Lyons et al. Nat Commun 10, 5120 (2019) “<a href="https://www.nature.com/articles/s41467-019-12515-9">Genome-wide association study of eosinophilic granulomatosis with polyangiitis reveals genomic loci stratified by ANCA status</a>”.</p>
</div>
<div id="data-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preprocessing" class="anchor"></a>Data preprocessing</h2>
<p>First we load the data and see what we have:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(lyons_egpa)
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(lyons_egpa,<span class="dt">n=</span><span class="dv">10</span>)[]</code></pre></div>
<pre><code>##     CHR      BP a1 a2        or        seb p.value
##  1:   1 1239953  A  G 0.9540296 0.13277346    0.52
##  2:   1 1744812  A  G 1.1077122 0.21219596    0.52
##  3:   1 2513216  C  T 1.0773496 0.05533328    0.59
##  4:   1 2526746  G  A 0.9825316 0.05751922    0.86
##  5:   1 2709164  C  A 1.0052302 0.05851229    0.74
##  6:   1 4698559  C  T 0.9576941 0.07558322    0.44
##  7:   1 5723376  G  A 0.9942471 0.05814479    0.71
##  8:   1 6548056  G  A 0.9044554 0.08667750    0.17
##  9:   1 8180210  G  A 1.0756336 0.07048286    0.33
## 10:   1 8469789  C  T 0.9891649 0.06197632    0.81</code></pre>
<p>Next let’s load in the basis as used in Burren et al. whilst many objects are loaded for data preprocessing we are most interested in the manifest file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(burren_imd_<span class="dv">13</span>_traits)
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(SNP.manifest,<span class="dt">n=</span><span class="dv">10</span>)[]</code></pre></div>
<pre><code>##              pid ref_a1 ref_a2 ref_a1.af ld.block chr
##  1: 10:101291593      T      C  0.480825     1984  10
##  2: 10:112186148      C      T  0.648373     1992  10
##  3: 10:115481018      C      T  0.744380     1997  10
##  4: 10:131562993      G      T  0.862206     2029  10
##  5:  10:27177245      A      G  0.865380     1913  10
##  6:  10:30802799      A      C  0.752050     1918  10
##  7:  10:33426147      T      C  0.616504     1920  10
##  8:  10:35359619      A      G  0.658291     1923  10
##  9:  10:59893930      C      T  0.695451     1940  10
## 10:   10:6087295      G      T  0.770167     1875  10</code></pre>
<p>The first job is to work out the intersect between the egpa data and the SNPs that we wish to include in the basis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#first add a pid column to the egpa dataset</span>
lyons_egpa[,pid<span class="op">:</span><span class="er">=</span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(CHR,BP,<span class="dt">sep=</span><span class="st">':'</span>)]
<span class="co"># check to see whether there are manifest SNPs missing</span>
M &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(SNP.manifest,lyons_egpa,<span class="dt">by=</span><span class="st">'pid'</span>)
<span class="co"># in this case there are no missing values.</span>
M[<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(or),]</code></pre></div>
<pre><code>## Empty data.table (0 rows and 13 cols): pid,ref_a1,ref_a2,ref_a1.af,ld.block,chr...</code></pre>
<p>In this particular example there are are no missing values, if you find that over 5% of values are missing you might wish to consider imputing missing values either by going back and imputing your GWAS or using a tool such as <a href="https://github.com/zkutalik/ssimp_software">ssimp</a>. The next task is to harmonise OR so that both basis and EGPA datasets are with respect to the same allele. Please note that for the lyons_egpa dataset allele 1 (a1) is the effect allele, for your dataset this might be different in which case you will need to flip when alleles match the manifest file (rather than when they are discordant as is shown below).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">M[,flip_allele<span class="op">:</span><span class="er">=</span><span class="ot">NA</span>][]</code></pre></div>
<pre><code>##               pid ref_a1 ref_a2 ref_a1.af ld.block chr CHR        BP a1 a2
##   1: 10:101291593      T      C  0.480825     1984  10  10 101291593  T  C
##   2: 10:112186148      C      T  0.648373     1992  10  10 112186148  C  T
##   3: 10:115481018      C      T  0.744380     1997  10  10 115481018  T  C
##   4: 10:131562993      G      T  0.862206     2029  10  10 131562993  G  T
##   5:  10:27177245      A      G  0.865380     1913  10  10  27177245  A  G
##  ---                                                                      
## 479:    9:5455732      T      G  0.747554     1716   9   9   5455732  G  T
## 480:    9:6161686      C      T  0.574054     1717   9   9   6161686  C  T
## 481:    9:6163823      G      A  0.573922     1717   9   9   6163823  G  A
## 482:   9:92321289      A      G  0.745041     1794   9   9  92321289  A  G
## 483:   9:93563536      A      G  0.801111     1796   9   9  93563536  A  G
##             or        seb p.value flip_allele
##   1: 0.8596157 0.05486505  0.0036          NA
##   2: 1.0434465 0.05911730  0.6000          NA
##   3: 0.9997357 0.06289752  0.7800          NA
##   4: 0.9845815 0.08011694  0.9800          NA
##   5: 1.0164064 0.08043653  0.8200          NA
##  ---                                         
## 479: 1.0846591 0.06255629  0.3200          NA
## 480: 0.9448833 0.05638120  0.2900          NA
## 481: 0.9445537 0.05623896  0.2900          NA
## 482: 1.0734127 0.06206261  0.5500          NA
## 483: 0.9686060 0.06904879  0.7400          NA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## in this case if we need do nothing as these are already aligned
M[ref_a1<span class="op">==</span>a1 <span class="op">&amp;</span><span class="st"> </span>ref_a2<span class="op">==</span>a2,flip_allele<span class="op">:</span><span class="er">=</span><span class="ot">FALSE</span>]
## in this case the effect allele is the wrong way around and so we need to flip
M[ref_a1<span class="op">==</span>a2 <span class="op">&amp;</span><span class="st"> </span>ref_a2<span class="op">==</span>a1,flip_allele<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>][,<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'a1'</span>,<span class="st">'a2'</span>,<span class="st">'or'</span>)<span class="op">:</span><span class="er">=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(a2,a1,<span class="dv">1</span><span class="op">/</span>or)][]</code></pre></div>
<pre><code>##               pid ref_a1 ref_a2 ref_a1.af ld.block chr CHR        BP a1 a2
##   1: 10:101291593      T      C  0.480825     1984  10  10 101291593  C  T
##   2: 10:112186148      C      T  0.648373     1992  10  10 112186148  T  C
##   3: 10:115481018      C      T  0.744380     1997  10  10 115481018  C  T
##   4: 10:131562993      G      T  0.862206     2029  10  10 131562993  T  G
##   5:  10:27177245      A      G  0.865380     1913  10  10  27177245  G  A
##  ---                                                                      
## 479:    9:5455732      T      G  0.747554     1716   9   9   5455732  T  G
## 480:    9:6161686      C      T  0.574054     1717   9   9   6161686  T  C
## 481:    9:6163823      G      A  0.573922     1717   9   9   6163823  A  G
## 482:   9:92321289      A      G  0.745041     1794   9   9  92321289  G  A
## 483:   9:93563536      A      G  0.801111     1796   9   9  93563536  G  A
##             or        seb p.value flip_allele
##   1: 1.1633106 0.05486505  0.0036       FALSE
##   2: 0.9583625 0.05911730  0.6000       FALSE
##   3: 1.0002644 0.06289752  0.7800        TRUE
##   4: 1.0156599 0.08011694  0.9800       FALSE
##   5: 0.9838584 0.08043653  0.8200       FALSE
##  ---                                         
## 479: 0.9219487 0.06255629  0.3200        TRUE
## 480: 1.0583317 0.05638120  0.2900       FALSE
## 481: 1.0587010 0.05623896  0.2900       FALSE
## 482: 0.9316081 0.06206261  0.5500       FALSE
## 483: 1.0324116 0.06904879  0.7400       FALSE</code></pre>
<p>In this simple case the alleles were easy to align however in you dataset things will probably be different and it is worth considering using something such as annotSnpStats that has more sophisticated routines for aligning alleles. Here is some example code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(annotSnpStats) <span class="co">#see https://github.com/chr1swallace/annotSnpStats</span>
alleles &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">pid=</span>M<span class="op">$</span>pid,<span class="dt">al.x =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(M<span class="op">$</span>ref_a1,M<span class="op">$</span>ref_a2,<span class="dt">sep=</span><span class="st">'/'</span>),<span class="dt">al.y=</span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(M<span class="op">$</span>a1,M<span class="op">$</span>a2,<span class="dt">sep=</span><span class="st">'/'</span>))
align.class &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">'match'</span>,<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(alleles))
idx&lt;-<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(alleles<span class="op">$</span>al.x<span class="op">!=</span>alleles<span class="op">$</span>al.y)
x.alleles &lt;-<span class="st"> </span>alleles[idx,]<span class="op">$</span>al.x
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(x.alleles)&lt;-alleles[idx,]<span class="op">$</span>pid
y.alleles &lt;-<span class="st">  </span>alleles[idx,]<span class="op">$</span>al.y
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(y.alleles)&lt;-<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(x.alleles)
align.class[idx] &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/annotSnpStats/man/g.class.html">g.class</a></span>(x.alleles,y.alleles)
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(align.class))
alleles[,g.class<span class="op">:</span><span class="er">=</span>align.class]
idx&lt;-<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(alleles<span class="op">$</span>g.class<span class="op">==</span><span class="st">'impossible'</span>)
<span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(idx) <span class="op">&gt;</span><span class="dv">0</span>){
    M &lt;-<span class="st"> </span>M[<span class="op">-</span>idx,]
    alleles &lt;-<span class="st"> </span>alleles[<span class="op">-</span>idx,]
}
M &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(M,alleles[,.(pid,g.class)],<span class="dt">by=</span><span class="st">'pid'</span>,<span class="dt">all.x=</span><span class="ot">TRUE</span>)
M &lt;-<span class="st"> </span>M[<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(pid),]
M &lt;-<span class="st"> </span>M[g.class<span class="op">==</span><span class="st">'match'</span>,or<span class="op">:</span><span class="er">=</span><span class="dv">1</span><span class="op">/</span>or]</code></pre></div>
</div>
<div id="projecting-onto-the-basis" class="section level2">
<h2 class="hasAnchor">
<a href="#projecting-onto-the-basis" class="anchor"></a>Projecting onto the basis</h2>
<p>We convert odds ratio estimates to the log odds ratio scale and use the cupcake function <em>project_sparse</em> to project the dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">proj.dt &lt;-<span class="st"> </span>cupcake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cupcake/man/project_sparse.html">project_sparse</a></span>(<span class="dt">beta=</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(M<span class="op">$</span>or),<span class="dt">seb=</span>M<span class="op">$</span>seb,<span class="dt">pids=</span>M<span class="op">$</span>pid)[,trait<span class="op">:</span><span class="er">=</span><span class="st">'EGPA'</span>][]</code></pre></div>
<pre><code>## Warning in cupcake::project_sparse(beta = log(M$or), seb = M$seb, pids = M$pid):
## more than 5% sparse basis snps missing</code></pre>
<p>Overall we can see that this trait is significant as 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10}, 2.254018210^{-10} We can check to see if which of the components differ significantly from control as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">proj.dt[p<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">/</span>.N,][]</code></pre></div>
<pre><code>##      PC       proj     var.proj      delta    p.overall        z            p
## 1: PC12 0.04222382 5.534777e-05 0.02428538 2.254018e-10 3.264336 1.097211e-03
## 2: PC13 0.02394193 6.163415e-05 0.05441644 2.254018e-10 6.931375 4.167685e-12
##    trait
## 1:  EGPA
## 2:  EGPA</code></pre>
<p>PC13 seems the most interesting and we can plot this in context like so, and see that at least for this component EGPA looks most like Asthma.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SPARSE_BASIS_EXTDATA &lt;-<span class="st"> '../inst/extdata/sparse_imd_basis'</span>
SNP_MANIFEST_FILE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(SPARSE_BASIS_EXTDATA,<span class="st">'support/sparse_snp_manifest.RDS'</span>)
TRAIT_MANIFEST_FILE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(SPARSE_BASIS_EXTDATA,<span class="st">'support/sparse_trait_manifest.tab'</span>)
SHRINKAGE_FILE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(SPARSE_BASIS_EXTDATA,<span class="st">'support/13_trait_sparse_shrinkage.RDS'</span>)
BASIS_FILE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(SPARSE_BASIS_EXTDATA,<span class="st">'support/13_trait_sparse_basis.RDS'</span>)
GWAS_DATA_DIR &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(SPARSE_BASIS_EXTDATA,<span class="st">'/gwas_data/'</span>)
gwas.DT&lt;-<span class="kw"><a href="../reference/get_gwas_data.html">get_gwas_data</a></span>(TRAIT_MANIFEST_FILE,SNP_MANIFEST_FILE,GWAS_DATA_DIR)</code></pre></div>
<pre><code>## Processing VIT</code></pre>
<pre><code>## Processing UC</code></pre>
<pre><code>## Processing CD</code></pre>
<pre><code>## Processing RA</code></pre>
<pre><code>## Processing SLE</code></pre>
<pre><code>## Processing T1D</code></pre>
<pre><code>## Processing PSC</code></pre>
<pre><code>## Processing asthma</code></pre>
<pre><code>## Processing PBC</code></pre>
<pre><code>## Processing CEL</code></pre>
<pre><code>## Processing MS</code></pre>
<pre><code>## Adding reference snp manifest annotations</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## project on original basis traits to characterise PCs
gwas.DT[,<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'beta'</span>,<span class="st">'seb'</span>)<span class="op">:</span><span class="er">=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(or),<span class="dv">1</span>) ]
bt.proj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/split.html">split</a></span>(gwas.DT,gwas.DT<span class="op">$</span>trait),<span class="cf">function</span>(x){
    tt &lt;-<span class="st"> </span>x<span class="op">$</span>trait <span class="op">%&gt;%</span><span class="st"> </span>unique
    cupcake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cupcake/man/project_sparse.html">project_sparse</a></span>(<span class="dt">beta=</span>x<span class="op">$</span>beta,<span class="dt">seb=</span>x<span class="op">$</span>seb,<span class="dt">pids=</span>x<span class="op">$</span>pid)[,trait<span class="op">:</span><span class="er">=</span>tt]
}) <span class="op">%&gt;%</span><span class="st"> </span>rbindlist</code></pre></div>
<pre><code>## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing

## Warning in cupcake::project_sparse(beta = x$beta, seb = x$seb, pids = x$pid):
## more than 5% sparse basis snps missing</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">comb.dt &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(bt.proj[,.(PC,delta,<span class="dt">var.proj=</span><span class="dv">0</span>,trait)],proj.dt[,.(PC,delta,var.proj,trait)])
comb.dt[var.proj<span class="op">!=</span><span class="dv">0</span>,ci<span class="op">:</span><span class="er">=</span><span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span>(var.proj) <span class="op">*</span><span class="st"> </span><span class="fl">1.96</span>]
plot.DT &lt;-<span class="st"> </span>comb.dt[PC<span class="op">==</span><span class="st">'PC13'</span>,][<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(delta,<span class="dt">decreasing =</span> <span class="ot">TRUE</span>),]
idx &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(plot.DT<span class="op">$</span>ci))
cols &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">'black'</span>,<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(plot.DT))
cols[idx] &lt;-<span class="st"> 'red'</span>
{
    <span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(plot.DT,<span class="kw"><a href="https://rdrr.io/r/graphics/dotchart.html">dotchart</a></span>(delta,<span class="dt">labels=</span>trait,<span class="dt">xlim=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="fl">0.1</span>,<span class="fl">0.05</span>),<span class="dt">pch=</span><span class="dv">19</span>,
                          <span class="dt">main=</span><span class="st">'PC13'</span>,<span class="dt">xlab=</span><span class="st">"Delta PC score"</span>,
                          <span class="dt">col=</span>cols))
    ## add 95% confidence intervals
    <span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(plot.DT[idx,],<span class="kw"><a href="https://rdrr.io/r/graphics/arrows.html">arrows</a></span>(delta<span class="op">-</span>ci, idx, delta<span class="op">+</span>ci, idx, <span class="dt">length=</span><span class="fl">0.05</span>, <span class="dt">angle=</span><span class="dv">90</span>, <span class="dt">code=</span><span class="dv">3</span>,<span class="dt">col=</span><span class="st">'red'</span>))
    <span class="kw"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="dt">v=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">'red'</span>,<span class="dt">lty=</span><span class="dv">2</span>)
}</code></pre></div>
<p><img src="project-basis_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#data-preprocessing">Data preprocessing</a></li>
      <li><a href="#projecting-onto-the-basis">Projecting onto the basis</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Olly Burren, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
